core/types/calls.cc

```diff
@@ -680,6 +692,129 @@ DispatchResult dispatchCallSymbol(const GlobalState &gs, DispatchArgs args,
+            auto kwend = args.args.begin() + args.numPosArgs + numKwargs;
+
+            while (kwit != kwend) {
+                // if the key isn't a symbol literal, break out as this is not a valid keyword
+                auto &key = *kwit++;
+                auto *lit = cast_type<LiteralType>(key->type.get());
+                if (!lit || lit->literalKind != LiteralType::LiteralTypeKind::Symbol) {
+                    // TODO(trevor) add an error
+
+                    keys.clear();
+                    values.clear();
+                    break;
+                }
+
+                auto &val = *kwit++;
+                keys.emplace_back(key->type);
+                values.emplace_back(val->type);
+            }
+        }
+
+        // merge in the keyword splat argument if it's present
+        if (hasKwsplat) {
+            if (hasKwargs) {
+                auto &kwSplatArg = *(aend - 1);
+                auto kwSplatType = Types::approximate(gs, kwSplatArg->type, *constr);
+
+                if (auto *hash = cast_type<ShapeType>(kwSplatType.get())) {
+                    absl::c_copy(hash->keys, back_inserter(keys));
+                    absl::c_copy(hash->values, back_inserter(values));
+                    kwargs = make_type<ShapeType>(Types::hashOfUntyped(), move(keys), move(values));
+                    --aend;
+                } else {
+                    if (kwSplatType->isUntyped()) {
+                        // Allow an untyped arg to satisfy all kwargs
+                        --aend;
+                        kwargs = Types::untypedUntracked();
+                    } else if (kwSplatType->derivesFrom(gs, Symbols::Hash())) {
+                        --aend;
+                        if (auto e =
+                                gs.beginError(core::Loc(args.locs.file, args.locs.call), errors::Infer::UntypedSplat)) {
+                            e.setHeader("Passing a hash where the specific keys are unknown to a method taking keyword "
+                                        "arguments");
+                            e.addErrorSection(ErrorSection("Got " + kwSplatType->show(gs) + " originating from:",
+                                                           kwSplatArg->origins2Explanations(gs)));
+                            result.main.errors.emplace_back(e.build());
+                        }
+                        kwargs = Types::untypedUntracked();
+                    }
+                }
+
+                // Check to see if the keyword splat was a valid kwargs hash, and consume a positional argument if it
+                // was implicit.
+                if (implicitKwsplat && kwargs != nullptr) {
+                    --posArgs;
+                }
+            } else {
+                // This function doesn't take keyword arguments, so consume the kwsplat and mark all the keyword args as
+                // an untyped hash.
+                kwargs = Types::hashOfUntyped();
```

Comment:
Should this be the approximated type?

Also, add a comment about how this isn't a warning in ruby, and is fine.

Reply:
+1 it would be nice to try to build a ShapeType here.